### PR TITLE
Update syntax of How-to-extend-ballerina guide

### PR DESCRIPTION
**Pull request by rdhananjaya**
## Purpose
$subject

### CODE REVIEW

File: website/two-column-pages/docs/learn/how-to-extend-ballerina.md [hunk 1]
```diff
@@ -43,144 +41,117 @@ import ballerina/http;
...
+        self.basicClient = new(TWILIO_API_BASE_URL, config = twilioConfig.basicClientConfig);
+        self.authyClient = new(AUTHY_API_BASE_URL, config = twilioConfig.authyClientConfig);
+        self.accountSId = twilioConfig.accountSId;
+        self.xAuthyKey = twilioConfig.xAuthyKey;
+    }
 
-    public function getAccountDetails() returns (Account|error);
-};
+    # Initialize Twilio endpoint.
+    # + twilioConfig - Twilio configuraion
```
Comment (Shan1024) [inline website/two-column-pages/docs/learn/how-to-extend-ballerina.md]: Better to add an empty doc line between the description and the parameter to increase the readability. Applicable to the below places as well.

File: website/two-column-pages/docs/learn/how-to-extend-ballerina.md [hunk 2]
```diff
@@ -43,144 +41,117 @@ import ballerina/http;
...
-### Implement the `getCallerActions()` Function for the Connector
 
-The `getCallerActions()` function is called whenever the system needs to return an active connection to the end user developer for use in their code. We have already initialized the connection and it is saved within the TwilioClient object.
-
-```ballerina
-function TwilioClient::getCallerActions() returns TwilioConnector {
-    return self.twilioConnector;
+public function Client.init(TwilioConfiguration twilioConfig) {
+    http:AuthConfig authConfig = {
+                            scheme: http:BASIC_AUTH,
```
Comment (Shan1024) [inline website/two-column-pages/docs/learn/how-to-extend-ballerina.md]: Seems like there are too many leading spaces.

File: website/two-column-pages/docs/learn/how-to-extend-ballerina.md [hunk 3]
```diff
@@ -43,144 +41,117 @@ import ballerina/http;
...
+public remote function Client.getAccountDetails() returns Account|error {
+    string requestPath = TWILIO_ACCOUNTS_API + FORWARD_SLASH + self.accountSId + ACCOUNT_DETAILS;
+    var response = self.basicClient->get(requestPath);
     json jsonResponse = check parseResponseToJson(response);
     return mapJsonToAccount(jsonResponse);
 }
 ```
 
-And the `Account` record is also defined in the same file:
+And the `Account` record is defined in twilio_types.bal file:
```
Comment (Shan1024) [inline website/two-column-pages/docs/learn/how-to-extend-ballerina.md]:
```suggestion
And the `Account` record is defined in `twilio_types.bal` file:
```

File: website/two-column-pages/docs/learn/how-to-extend-ballerina.md [hunk 4]
```diff
@@ -14,13 +14,11 @@ To create a client connector, you:
 
 1. Create a Ballerina module in a Ballerina project.
 
-2. Create an object with an `init()` and `getCallerActions()` function.
+2. Create a `client` object representing the connector.
 
-3. Implement the `init()` function, which is called when the user instantiates an endpoint.
+3. Implement public remote functions to represent actions performed on remote endpoint.
```
[the remark this answers is not in the document]
Comment (praneesha) [reply website/two-column-pages/docs/learn/how-to-extend-ballerina.md]:
```suggestion
3. Implement `public` `remote` functions to represent actions performed on remote endpoints.
```
+1 to highlight.

File: website/two-column-pages/docs/learn/how-to-extend-ballerina.md [hunk 5]
```diff
@@ -14,13 +14,11 @@ To create a client connector, you:
...
 
-4. Implement the `getCallerActions()` function, which is called when the connection must be returned.
-
-5. Build the module and push it into a registry for usage by others.
+4. Build the module and push it into a registry for usage by others.
```
Comment (praneesha) [inline website/two-column-pages/docs/learn/how-to-extend-ballerina.md]:
```suggestion
4. Build the module and push it into a registry for others to use it.
```

File: website/two-column-pages/docs/learn/how-to-extend-ballerina.md [hunk 6]
```diff
@@ -43,144 +41,117 @@ import ballerina/http;
...
+    });
 
-    Account account = check twilioClient->getAccountDetails();
+    twilio:Account account = check twilioClient->getAccountDetails();
     io:println(account);
 }
 ```
 
-In this example, the endpoint is going to instantiate a `TwilioClient` object. This object takes an input parameter defined as a `clientConfig` object. The `clientConfig` object includes an `auth` object, which is part of the `http:Client` connector, which is distributed as part of the standard library.
+In this example, the endpoint is going to instantiate a `TwilioClient` object. Instantiating this requires `TwilioConfiguration` record instance to provide the authentication details.
```
Comment (praneesha) [inline website/two-column-pages/docs/learn/how-to-extend-ballerina.md]:
```suggestion
In this example, the endpoint instantiates a `TwilioClient` object. Instantiating this requires the `TwilioConfiguration` record instance to provide the authentication details.
```
Is saying "record instance" correct?

File: website/two-column-pages/docs/learn/how-to-extend-ballerina.md [hunk 7]
```diff
@@ -43,144 +41,117 @@ import ballerina/http;
...
 
-The Twilio connector then defines a custom function `getAccountDetails()`, which is called by the end user to interact with the endpoint. The module developer will also implement a `TwilioClient::init()` method, which will be called when the endpoint is instantiated. This method establishes the connection to Twilio.
+The Twilio connector then defines a custom function `getAccountDetails()`, which is called by the end user to interact with the endpoint. The module developer will also implement a `TwilioClient:init()` method, which will be called when the endpoint is instantiated. This method establishes the connection to Twilio.
```
Comment (praneesha) [inline website/two-column-pages/docs/learn/how-to-extend-ballerina.md]:
```suggestion
The Twilio connector then defines a custom function `getAccountDetails()`, which is called by the end user to interact with the endpoint. Also, you need to implement a `TwilioClient:init()` method, which will be called when the endpoint is instantiated. This method establishes the connection to Twilio.
```

File: website/two-column-pages/docs/learn/how-to-extend-ballerina.md [hunk 8]
```diff
@@ -43,144 +41,117 @@ import ballerina/http;
...
 
-### The TwilioClient Object
+### The Twilio:Client Object
 
-The connector is a data structure that is represented by an `object`.
+The connector is a data structure that is represented by an client `object`.
```
Comment (praneesha) [inline website/two-column-pages/docs/learn/how-to-extend-ballerina.md]:
```suggestion
The connector is a data structure that is represented by a `client` object.
```

File: website/two-column-pages/docs/learn/how-to-extend-ballerina.md [hunk 9]
```diff
@@ -43,144 +41,117 @@ import ballerina/http;
...
-    public function getAccountDetails() returns (Account|error);
-};
+    # Initialize Twilio endpoint.
+    # + twilioConfig - Twilio configuraion
+    public function init(TwilioConfiguration twilioConfig);
+    ...
 ```
 
-### Implement the `init()` Function for the Connector
+Client endpoint is similar to ordinary object, but unlike ordinary objects in client objects, the actions are performed on a remote network endpoint, hence the 'remote' keyword in method signature.
```
Comment (praneesha) [inline website/two-column-pages/docs/learn/how-to-extend-ballerina.md]:
```suggestion
A `client` endpoint is similar to an ordinary object. However, unlike in ordinary objects in `client` objects, the actions are performed on a remote network endpoint. Hence, the 'remote' keyword should be used in the method signature.
```

File: website/two-column-pages/docs/learn/how-to-extend-ballerina.md [hunk 10]
```diff
@@ -43,144 +41,117 @@ import ballerina/http;
...
 
-We now implement the method that will be called when an endpoint is instantiated. The function name is the object we created with `::init()` provided to reference that init method. The initialization method can accept a single parameter. When the end user calles the `endpoint` object, the record literals after the endpoint object will be mapped and passed in as the parameter to this method.
+Similar to other objects we take our configurations parameters such as API keys, as constructor parameter.
```
Comment (praneesha) [inline website/two-column-pages/docs/learn/how-to-extend-ballerina.md]:
```suggestion
Similar to other objects, `client` objects take their configurations parameters such as API keys as the values of the `constructor` parameter.
```

File: website/two-column-pages/docs/learn/how-to-extend-ballerina.md [hunk 11]
```diff
@@ -43,144 +41,117 @@ import ballerina/http;
...
+        self.init(twilioConfig);
+        self.basicClient = new(TWILIO_API_BASE_URL, config = twilioConfig.basicClientConfig);
+        self.authyClient = new(AUTHY_API_BASE_URL, config = twilioConfig.authyClientConfig);
+        self.accountSId = twilioConfig.accountSId;
+        self.xAuthyKey = twilioConfig.xAuthyKey;
     }
 
-    // We can reference the fields of the TwilioClient object
-    self.twilioConnector.accountSId = username;
+    # Initialize Twilio endpoint.
```
Comment (praneesha) [inline website/two-column-pages/docs/learn/how-to-extend-ballerina.md]:
```suggestion
    # Initialize the Twilio endpoint.
```

File: website/two-column-pages/docs/learn/how-to-extend-ballerina.md [hunk 12]
```diff
@@ -43,144 +41,117 @@ import ballerina/http;
...
+    # + twilioConfig - Twilio configuraion
```
Comment (praneesha) [inline website/two-column-pages/docs/learn/how-to-extend-ballerina.md]:
```suggestion
    # + twilioConfig - The Twilio configuraion
```

File: website/two-column-pages/docs/learn/how-to-extend-ballerina.md [hunk 13]
```diff
@@ -43,144 +41,117 @@ import ballerina/http;
...
+    twilioConfig.basicClientConfig.auth = authConfig;
 }
+
 ```
 
-### Implement Custom Functions For Endpoint Interaction
+### Implement Remote Functions For Endpoint Interaction
 
-While the `TwilioClient` object implements `init()` and `getCallerActions()`, if you want to expose custom actions for your end users to use against the connector, these are defined in the `TwilioConnector` object, which was initialized and stored as a reference to the `TwilioClient` object. You can add as many or as few custom functions to this object.
+When you want to expose custom actions for your end users to use against the connector, we define them in the `Twilio:Client` object as remote functions
```
Comment (praneesha) [inline website/two-column-pages/docs/learn/how-to-extend-ballerina.md]:
```suggestion
When you want to expose custom actions for your end users to use against the connector, define them in the `Twilio:Client` object as remote functions.
```

File: website/two-column-pages/docs/learn/how-to-extend-ballerina.md (hunk 3 above)
Comment (praneesha) [reply website/two-column-pages/docs/learn/how-to-extend-ballerina.md]:
```suggestion
The `Account` record is defined in the `twilio_types.bal` file:
```
Better not start sentences with "and".

File: website/two-column-pages/docs/learn/how-to-extend-ballerina.md [hunk 14]
```diff
@@ -411,8 +367,19 @@ In the `pom.xml`, add Ballerina's repository information.
 </repositories>
 ```
 
+##### Remove Some Unnecessary Files
+
+Remove the archetype generated `App.java` and `AppTest.java` files. They are not needed.
```
Comment (praneesha) [inline website/two-column-pages/docs/learn/how-to-extend-ballerina.md]:
```suggestion
Remove the archetype generated by the `App.java` and `AppTest.java` files. They are not needed.
```

File: website/two-column-pages/docs/learn/how-to-extend-ballerina.md [hunk 15]
```diff
@@ -411,8 +367,19 @@ In the `pom.xml`, add Ballerina's repository information.
...
+And then add an xml file named `spotbugs-exclude.xml` beside `pom.xml` and add the following.
```
Comment (praneesha) [inline website/two-column-pages/docs/learn/how-to-extend-ballerina.md]:
```suggestion
Then, add an XML file named `spotbugs-exclude.xml` together with the `pom.xml` and add the following.
```

File: website/two-column-pages/docs/learn/how-to-extend-ballerina.md [hunk 16]
```diff
@@ -43,144 +41,119 @@ import ballerina/http;
...
+    });
 
-    Account account = check twilioClient->getAccountDetails();
+    twilio:Account account = check twilioClient->getAccountDetails();
     io:println(account);
 }
 ```
 
-In this example, the endpoint is going to instantiate a `TwilioClient` object. This object takes an input parameter defined as a `clientConfig` object. The `clientConfig` object includes an `auth` object, which is part of the `http:Client` connector, which is distributed as part of the standard library.
+In this example, the endpoint instantiates a `TwilioClient` object. Instantiating this requires the `TwilioConfiguration` record instance to provide the authentication details.
```
Comment (MaryamZi) [inline website/two-column-pages/docs/learn/how-to-extend-ballerina.md]: I don't think we need to talk about endpoint now specifically.

```suggestion
In this example, a `TwilioClient` object is instantiated. Instantiating the client requires a `TwilioConfiguration` record to provide the authentication details.
```

File: website/two-column-pages/docs/learn/how-to-extend-ballerina.md [hunk 17]
```diff
@@ -43,144 +41,119 @@ import ballerina/http;
...
 
-The Twilio connector then defines a custom function `getAccountDetails()`, which is called by the end user to interact with the endpoint. The module developer will also implement a `TwilioClient::init()` method, which will be called when the endpoint is instantiated. This method establishes the connection to Twilio.
+The Twilio connector then defines a custom function `getAccountDetails()`, which is called by the end user to interact with the endpoint. Also, you need to implement a `TwilioClient:init()` method, which will be called when the endpoint is instantiated. This method establishes the connection to Twilio.
```
Comment (MaryamZi) [inline website/two-column-pages/docs/learn/how-to-extend-ballerina.md]:
```suggestion
The Twilio connector defines a remote function `getAccountDetails()`, which is called by the end user to interact with the endpoint. 

You need to implement the `TwilioClient.__init()` method, which will be called when the endpoint is instantiated. This method establishes the connection to Twilio.
```

**Comment by MaryamZi** [comment]
Shouldn't https://github.com/wso2-ballerina/module-twilio/blob/master/guide/twilio_sample.bal that is linked here also be updated?

**Comment by rdhananjaya** [comment]
@MaryamZi This should update it. https://github.com/wso2-ballerina/module-twilio/pull/48

File: website/two-column-pages/docs/learn/how-to-extend-ballerina.md [hunk 18]
```diff
@@ -43,144 +41,121 @@ import ballerina/http;
...
+        self.accountSId = twilioConfig.accountSId;
+        self.xAuthyKey = twilioConfig.xAuthyKey;
+    }
 
+    # Initialize Twilio endpoint.
+    # + twilioConfig - Twilio configuraion
 
-    public function getAccountDetails() returns (Account|error);
-};
+    public function init(TwilioConfiguration twilioConfig);
```
Comment (MaryamZi) [inline website/two-column-pages/docs/learn/how-to-extend-ballerina.md]: Shall we not expose this? The relevant function is __init. I believe this was leftover from the previous implementation.

File: website/two-column-pages/docs/learn/how-to-extend-ballerina.md [hunk 19]
```diff
@@ -43,144 +41,121 @@ import ballerina/http;
...
+    ...
 ```
 
-### Implement the `init()` Function for the Connector
+A `client` endpoint is similar to an ordinary object. However, unlike in ordinary objects in `client` objects, the actions are performed on a remote network endpoint. Hence, the 'remote' keyword should be used in the method signature.
```
Comment (MaryamZi) [inline website/two-column-pages/docs/learn/how-to-extend-ballerina.md]:
```suggestion
A `client` object is similar to an ordinary object. However, unlike in ordinary objects in `client` objects, the actions are performed on a remote network endpoint. Hence, the 'remote' keyword should be used in the method signature.
```

File: website/two-column-pages/docs/learn/how-to-extend-ballerina.md [hunk 20]
```diff
@@ -43,144 +41,117 @@ import ballerina/http;
 import wso2/twilio;
 
 function main(string... args) {
```
Comment (MaryamZi) [inline website/two-column-pages/docs/learn/how-to-extend-ballerina.md]:
```suggestion
public function main(string... args) {
```

File: website/two-column-pages/docs/learn/how-to-extend-ballerina.md [hunk 21]
```diff
@@ -43,144 +41,117 @@ import ballerina/http;
...
+    # Initialize the Twilio endpoint.
+    # + twilioConfig - The Twilio configuration
 
-    // Calling initialize of the embedded http client object
-    self.twilioConnector.client.init(config.clientConfig);
-}
-```
-
-### Implement the `getCallerActions()` Function for the Connector
+    public function init(TwilioConfiguration twilioConfig);
```
Comment (MaryamZi) [inline website/two-column-pages/docs/learn/how-to-extend-ballerina.md]: shouldn't we remove this too?

Comment (MaryamZi) [reply website/two-column-pages/docs/learn/how-to-extend-ballerina.md]: or maybe not, since this is the complete file.

Comment (rdhananjaya) [reply website/two-column-pages/docs/learn/how-to-extend-ballerina.md]: Yes, let's let it be, as it's just a function.

**Pull request merged**